lib/listen/fsm.rb

```diff
@@ -1,131 +1,128 @@
+      def start_state(new_start_state = nil)
+        if new_start_state
+          new_start_state.is_a?(Symbol) or raise ArgumentError, "state name must be a Symbol (got #{new_start_state.inspect})"
+          @start_state = new_start_state
         else
-          defined?(@default_state) ? @default_state : DEFAULT_STATE
+          defined?(@start_state) ? @start_state : START_STATE
         end
       end
 
-      # Obtain the valid states for this FSM
+      # The valid states for this FSM, as a hash with state name symbols as keys and State objects as values.
       def states
         @states ||= {}
       end
 
-      # Declare an FSM state and optionally provide a callback block to fire
+      # Declare an FSM state and optionally provide a callback block to fire on state entry
       # Options:
       # * to: a state or array of states this state can transition to
-      def state(*args, &block)
-        if args.last.is_a? Hash
-          # Stringify keys :/
-          options = args.pop.each_with_object({}) { |(k, v), h| h[k.to_s] = v }
-        else
-          options = {}
-        end
-
-        args.each do |name|
-          name = name.to_sym
-          default_state name if options['default']
-          states[name] = State.new(name, options['to'], &block)
-        end
+      def state(state_name, to: nil, &block)
+        state_name.is_a?(Symbol) or raise ArgumentError, "state name must be a Symbol (got #{state_name.inspect})"
+        states[state_name] = State.new(state_name, to, &block)
       end
     end
 
-    # Be kind and call super if you must redefine initialize
+    # Note: you must call super() from including classes so this code will run
     def initialize
-      @state = self.class.default_state
+      @state = self.class.start_state
+      @mutex = ::Mutex.new
+      @state_changed = ::ConditionVariable.new
+      super
     end
 
-    # Obtain the current state of the FSM
+    # Current state of the FSM, stored as a symbol
     attr_reader :state
 
-    def transition(state_name)
-      new_state = validate_and_sanitize_new_state(state_name)
-      return unless new_state
-      transition_with_callbacks!(new_state)
+    def transition(new_state_name)
+      new_state_name.is_a?(Symbol) or raise ArgumentError, "state name must be a Symbol (got #{new_state_name.inspect})"
+      if (new_state = validate_and_sanitize_new_state(new_state_name))
+        transition_with_callbacks!(new_state)
+      end
     end
 
-    # Immediate state transition with no checks, or callbacks. "Dangerous!"
-    def transition!(state_name)
-      @state = state_name
+    # Low-level, immediate state transition with no checks or callbacks.
+    def transition!(new_state_name)
+      new_state_name.is_a?(Symbol) or raise ArgumentError, "state name must be a Symbol (got #{new_state_name.inspect})"
+      @mutex or raise ArgumentError, "FSM not initialized. You must call super() from initialize!"
```

Comment:
This seems like an overly specific issue to check for - I can imagine some code was not working correctly, but in that case `@mutex.synchronize` will blow up with nil error... do we need to manually check it?

Reply:
Understood.

Well, I don't really want to make the code any more complicated than it already is.

If the code is not implemented correctly (missing call to `super`) it's a mechanical problem - i.e. the code needs to be fixed. This isn't a transient run-time issue. The current implementation is okay, but it's also very explicit in a non-canonical way.

I don't think we should be adding extra instance variables to enforce initialisation.

Reply:
I realized we had to move away from  `super()`. There's a reason I've never seen mixins initialized that way before: it precludes any base class initialization, since the mixin would have to do that for you (by calling `super` itself), but would have no way to know the base class's `initialize` signature!

Instead I went with an explicitly-named method `initialize_fsm`. And stored a variable to remember if that didn't happen. And I addressed the longstanding comment in there to make the FSM methods private. And I added a spec for the FSM module and tested all of this. I think it's all much cleaner now! LMK what you think.